### PR TITLE
feat!: implement magento url rewrites

### PR DESCRIPTION
**Pull request by sequensucks**
## Description
Use a magento route graphql query and implement seo-friendly urls for product, category and cms
pages

BREAKING CHANGE: Change a urls pattern for product, category, cms pages

## Related Issue
re #181

## How Has This Been Tested?
Check all urls for product, category, cms pages 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by Frodigo — CHANGES_REQUESTED** [review]
Good job at all, thanks! 

I added some initial comments. Please take a look.

File: packages/api-client/src/api/route/route.ts [hunk 1]
```diff
@@ -4,12 +4,43 @@ import gql from 'graphql-tag';
 export default gql`
   query route($url: String!) {
     route(url: $url) {
-      relative_url
-      redirect_code
-      type
-      ... on CategoryTree {
+      __typename
+      ... on SimpleProduct {
+        sku
```
[the remark this answers is not in the document]
Comment (sequensucks) [reply packages/api-client/src/api/route/route.ts:9]: Sorry, i use this query on my project and forgot to remove unused fields. Fixed

File: packages/theme/composables/index.ts [hunk 2]
```diff
@@ -36,7 +36,6 @@ export * from './useUiHelpers';
 export * from './useUiNotification';
 export * from './useUiState';
 export * from '../modules/catalog/product/composables/useUpsellProducts';
-export * from './useUrlResolver';
```
[the remark this answers is not in the document]
Comment (sequensucks) [reply packages/theme/composables/index.ts]: Restored it.

File: packages/theme/middleware/url-resolver.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+  const clearUrl = path.replace(/[a-z]+\/[cp|]\//gi, '').replace(`/${context.i18n.locale}`, '');
+
+  Logger.debug('middleware/url-resolver', clearUrl);
+
+  const { data } = await context.app.$vsf.$magento.api.route(clearUrl);
+
+  Logger.debug('middleware/url-resolver/result', { data });
+
+  // @ts-ignore
```
[the remark this answers is not in the document]
Comment (sequensucks) [reply packages/theme/middleware/url-resolver.ts]: Removed a ts-ignore. There is no error here.

File: packages/theme/middleware/url-resolver.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import { Middleware } from '@nuxt/types';
+import { usePageStore } from '~/modules/catalog/store/page';
+import { Logger } from '~/helpers/logger';
+import { RoutableInterface } from '~/modules/GraphQL/types';
+
+const middleware : Middleware = async (context) => {
```
[the remark this answers is not in the document]
Comment (sequensucks) [reply packages/theme/middleware/url-resolver.ts]: added a ~/middleware/__tests__/url-resolver.spec.js

File: packages/theme/modules/catalog/getters/pageGetters.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+export interface PageGetters {
```
[the remark this answers is not in the document]
Comment (sequensucks) [reply packages/theme/modules/catalog/getters/pageGetters.ts]: Removed this method. We don't need it.
I replaced __typename with type and renamed pages components. After these changes we don't need an extra method.

File: packages/theme/modules/catalog/product/getters/productGetters.ts [hunk 6]
```diff
@@ -234,6 +235,11 @@ const getCategoryBreadcrumbs = (category: CategoryInterface): Breadcrumb[] => {
   return breadcrumbs;
 };
 
+export const getProductPath = (product: ProductInterface) => {
```
[the remark this answers is not in the document]
Comment (sequensucks) [reply packages/theme/modules/catalog/product/getters/productGetters.ts]: Moved to useProduct

File: packages/theme/middleware/url-resolver.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Frodigo) [reply packages/theme/middleware/url-resolver.ts]: Thanks!

File: packages/theme/modules/catalog/product/getters/productGetters.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Frodigo) [reply packages/theme/modules/catalog/product/getters/productGetters.ts]: great, thanks!

File: packages/api-client/src/api/route/route.ts [hunk 7]
```diff
@@ -4,12 +4,30 @@ import gql from 'graphql-tag';
 export default gql`
   query route($url: String!) {
     route(url: $url) {
-      relative_url
-      redirect_code
       type
+      ... on SimpleProduct {
```
Comment (sethidden) [inline packages/api-client/src/api/route/route.ts:8]: All of these product variants inherit for ProductInterface. The top-level type that creates "uid" is `ProductInterface`, not the child `SimpleProduct`, `GroupedProduct` itp.

So theoretically the duplicated `sku`s could be remove and just 
```
query aaa {
  route(url: "something") {
    ... on ProductInterface {
      uid
    }
  }
}
```
would work

Comment (sequensucks) [reply packages/api-client/src/api/route/route.ts:8]: oh, thank you! I tested it and it works perfect! I will update a query.

Comment (sequensucks) [reply packages/api-client/src/api/route/route.ts]: updated the query

**Comment by sequensucks** [comment]
@Frodigo ofc! 
- moved `page.vue`, `cms.vue` pages to packages/theme/pages
- and moved `~/modules/catalog/store/page.ts` to `~/stores/page.ts`

**Review by Frodigo — APPROVED** [review]

**Pull request merged**